bpf: fib: fix issues with L2 resolution

julianwiedmann opened the pull request:
Clarify in which scenarios the fallback to the neigh-map makes sense.

Then fix a recent regression where some code paths are no longer able to obtain a L2 resolution on BPF redirect. This would for instance affect the forwarding of LBed requests to remote backends, when running in XDP or on a pre-5.10 kernel.

borkmann commented on bpf/lib/fib.h:
```diff
@@ -66,8 +66,8 @@ static __always_inline bool fib_ok(int ret)
-		const struct bpf_fib_lookup_padded *fib_params, __s8 *fib_ret,
-		int *oif)
+		const struct bpf_fib_lookup_padded *fib_params,
+		bool lookup_done, __s8 *fib_ret, int *oif)
```
Commenting here given GH doesn't allow to comment on the commit message itself. Could you elaborate on the XDP situation? We do install managed neigh entries for all cluster nodes, so that for XDP we have the backends readily available. How does the scenario/issue you are fixing apply there?

julianwiedmann replied:
Hm, I was only concerned about the reply path.

Yes, the forward path doesn't look good either. Say `tail_nodeport_nat_egress_ipv4()` - it calls `fib_redirect()`. But why would it **ever** be allowed to skip the L2 resolution? It won't find the worker node's MAC address in the neighbour map. So this seems like another bug - we need better instruction by the caller whether the FIB lookup is required, and whether a fallback to neighbour map should even be attempted (it doesn't make sense for quite a few paths ...)

borkmann replied:
Hm, and what does the route lookup say? Does it point to a gateway as next hop which would be in the neigh table of the kernel? Or is the forward path expecting the backend in same L2 and somehow the resolution failed?

julianwiedmann replied:
>Yes, the forward path doesn't look good either. Say tail_nodeport_nat_egress_ipv4() - it calls fib_redirect(). But why would it ever be allowed to skip the L2 resolution? It won't find the worker node's MAC address in the neighbour map. So this seems like another bug - we need better instruction by the caller whether the FIB lookup is required, and whether a fallback to neighbour map should even be attempted (it doesn't make sense for quite a few paths ...)

This is addressed by the second patch now (ensuring that we perform a FIB lookup when the neigh-resolver is not available).

julianwiedmann committed:
bpf: fib: require opt-in for neighour map fallback in fib_do_redirect()

The neighbour map is populated by the inbound nodeport path, and used to
cache the client's MAC address. Therefore it only makes sense to use this
fallback in the LB's reply path.

Opt-out from using it in
- the LB NAT forward path
- the LB DSR forward path
- the outbound EgressGW paths
- bpf_lxc's reply path, as that's only used with ENABLE_HOST_ROUTING and
  thus can always use the neigh-resolver.

Note that callers which can't use the neigh-map will need *some* sort of
toleration for failed L2 resolution / DROP_NO_FIB result.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: fib: fix DMAC rewrite with ENABLE_SKIP_FIB

A recent FIB refactor introduced a bug, where fib_redirect*() no longer
performs a FIB lookup if ENABLE_SKIP_FIB is set.

But for configs without neigh-resolver, some code paths (that can't fall
back to the neigh map) strictly require this FIB lookup to obtain the
next-hop's MAC address.

Fix things by reintroducing the FIB lookup when neigh_resolver_available()
returns false.

Fixes: e30e18b646f6 ("bpf,fib: use fib_do_redirect in fib_redirect")
Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: fib: refactor fib_do_redirect()

Clarify the different paths of L2 resolution:
1. when the neigh-resolver is available, always use it. Forward the
   next-hop info from a preceding FIB lookup where available.
2. otherwise fallback to the neigh map, for callers that have opted in.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

ldelossa commented on bpf/lib/fib.h:
```diff
@@ -124,36 +121,39 @@ fib_do_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
 		 * prefer that over the BPF neighbor map since the latter
 		 * might be less accurate in some asymmetric corner cases.
 		 */
 		if (neigh_resolver_available()) {
```
I'm a bit confused about this if statement. You no longer value the 'use_neigh_map' boolean. If we want to move toward being "more explicit" oh what the function is doing based on boolean arguments, doesn't this throw an override on the expected behavior when the function is called?

julianwiedmann replied:
It's intended behaviour - when redirect_neigh() is available, we always want to prefer it over the neigh map.

I'll add proper docs for all the parameters. We can also call it `allow_neigh_map_fallback` or such.

ldelossa replied:
Yeah maybe that variable name update would suggest whats happening better.

julianwiedmann committed:
bpf: nodeport: opt-out from neighbour map when XDP-forwarding via tunnel

When XDP manually builds the tunnel headers and forwards to a remote node,
it makes no sense to rely on the neighbour map for L2 resolution.

We have to trust that the agent installs managed neigh entries for all
other nodes, and thus the FIB lookup will always return a L2 resolution.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

ldelossa reviewed (approved):
Read through these changes. 

They look good to me. Think we still have a bit of refactoring to go as we talked a bit about out of band, but this fixes the issue at hand.

jschwinger233 commented on bpf/lib/fib.h:
```diff
@@ -127,43 +123,37 @@ fib_do_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
+				 * inbound client traffic on the load-balancer where we
+				 * know that replies need to go back to them.
+				 */
+				dmac = fib_params->l.family == AF_INET ?
```
Do we need to check `if (fib_params)` here?

julianwiedmann replied:
I think that's something we'll need to investigate in a follow-on PR.

For users of `fib_redirect_v*()`, it gets all handled transparently - if there's no neigh-resolver available, they will always prepare + pass a `fib_params` struct.

"Expert" users that call `fib_do_redirect()` need to have some understanding what's needed for the fall back to the neigh-map. It's not easy to make that fool-proof. They could easily pass a 0-initialized `fib_params` - it would satisfy the suggested check and would satisfy the verifier, but still not help the neigh-map lookup.

jschwinger233 reviewed (approved):
Thanks! Just a minor question regarding `fib_params`.
